File: packages/sdk-core/src/__tests__/dsn.test.ts
import { describe, expect, it } from 'vitest';

import { InvalidDsnError, parseDsn } from '../dsn.js';

describe('parseDsn', () => {
  it('parses a production DSN and picks https transport for non-loopback hosts', () => {
    const dsn = parseDsn('arguslog://abc123@ingest.arguslog.io/api/42');
    expect(dsn).toEqual({
      protocol: 'https',
      publicKey: 'abc123',
      host: 'ingest.arguslog.io',
      projectId: '42',
      ingestUrl: 'https://ingest.arguslog.io/api/42/events',
    });
  });

  it('picks http transport for localhost', () => {
    const dsn = parseDsn('arguslog://key@localhost:8080/api/1');
    expect(dsn.protocol).toBe('http');
    expect(dsn.ingestUrl).toBe('http://localhost:8080/api/1/events');
  });

  it('picks http transport for 127.0.0.1', () => {
    expect(parseDsn('arguslog://k@127.0.0.1:8080/api/1').protocol).toBe('http');
  });

  // RFC1918 + just-outside-the-range cases live in scripts/dsn-test-fixtures.json so the
  // TS, Java and Python SDKs all run identical assertions; see dsn.fixtures.test.ts.

  it.each([
    'not-a-dsn',
    '',
    // missing public key
    'arguslog://@arguslog.io/api/42',
    // wrong scheme
    'https://key@arguslog.io/api/42',
    // wrong path prefix
    'arguslog://key@arguslog.io/42',
    // empty project id
    'arguslog://key@arguslog.io/api/',
  ])('rejects invalid DSN: %s', (bad) => {
    expect(() => parseDsn(bad)).toThrow(InvalidDsnError);
  });
});

File: java-sdk/src/test/java/org/arguslog/sdk/DsnTest.java
package org.arguslog.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;

import org.junit.jupiter.api.Test;

class DsnTest {

  @Test
  void parsesProductionDsnAsHttps() {
    Dsn dsn = Dsn.parse("arguslog://abc123@ingest.arguslog.io/api/42");
    assertThat(dsn.publicKey()).isEqualTo("abc123");
    assertThat(dsn.host()).isEqualTo("ingest.arguslog.io");
    assertThat(dsn.projectId()).isEqualTo("42");
    assertThat(dsn.scheme()).isEqualTo("https");
    assertThat(dsn.ingestUrl()).isEqualTo("https://ingest.arguslog.io/api/42/events");
  }

  @Test
  void parsesLoopbackDsnAsHttp() {
    Dsn dsn = Dsn.parse("arguslog://k@localhost:8080/api/1");
    assertThat(dsn.scheme()).isEqualTo("http");
    assertThat(dsn.ingestUrl()).isEqualTo("http://localhost:8080/api/1/events");
  }

  @Test
  void parses127LoopbackAsHttp() {
    Dsn dsn = Dsn.parse("arguslog://k@127.0.0.1:9000/api/7");
    assertThat(dsn.scheme()).isEqualTo("http");
    assertThat(dsn.ingestUrl()).isEqualTo("http://127.0.0.1:9000/api/7/events");
  }

  // RFC1918 + just-outside-the-range cases live in scripts/dsn-test-fixtures.json so all
  // three SDKs run identical assertions; see DsnFixturesTest.

  @Test
  void rejectsLegacyHttpsScheme() {
    assertThatIllegalArgumentException().isThrownBy(() -> Dsn.parse("https://k@host/1"));
  }

  @Test
  void rejectsMissingApiSegment() {
    assertThatIllegalArgumentException().isThrownBy(() -> Dsn.parse("arguslog://k@host/1"));
  }

  @Test
  void rejectsMissingProjectId() {
    assertThatIllegalArgumentException().isThrownBy(() -> Dsn.parse("arguslog://k@host/api/"));
  }

  @Test
  void rejectsMissingPublicKey() {
    assertThatIllegalArgumentException().isThrownBy(() -> Dsn.parse("arguslog://@host/api/1"));
  }
}
